crates/ra_ide/src/completion/complete_trait_impl.rs

```diff
@@ -0,0 +1,356 @@
+use crate::completion::{
+    CompletionContext, CompletionItem, CompletionItemKind, CompletionKind, Completions,
+};
+
+use ra_syntax::{ast::{self, edit}, AstNode, SyntaxKind, TextRange};
+use hir::{self, Docs, HasSource};
+
+use ra_assists::utils::get_missing_impl_items;
+
+/// Analyzes the specified `CompletionContext` and provides magic completions
+/// if the context falls within a `impl Trait for` block.
+/// 
+/// # Completion Activation
+/// The completion will activate when a user begins to type a function 
+/// definition, an associated type, or an associated constant.
+/// 
+/// ### Functions
+/// ```ignore
+/// trait SomeTrait {
+///     fn foo(&self);
+/// }
+/// 
+/// impl SomeTrait for () {
+///     fn <|>
+/// }
+/// ```
+/// 
+/// ### Associated Types
+/// ```ignore
+/// trait SomeTrait {
+///     type SomeType;
+/// }
+/// 
+/// impl SomeTrait for () {
+///     type <|>
+/// }
+/// ```
+/// 
+/// ### Associated Constants
+/// ```ignore
+/// trait SomeTrait {
+///     const SOME_CONST: u16;
+/// }
+/// 
+/// impl SomeTrait for () {
+///     const <|>
+/// }
+/// ```
+pub(crate) fn complete_trait_impl(acc: &mut Completions, ctx: &CompletionContext) {
+
+    // it is possible to have a parent `fn` and `impl` block. Ignore completion 
+    // attempts from within a `fn` block.
+    if ctx.function_syntax.is_some() {
+        return;
+    }
+
+    if let Some(ref impl_block) = ctx.impl_block {
+        for item in get_missing_impl_items(ctx.db, &ctx.analyzer, impl_block) {
+            match item {
+                hir::AssocItem::Function(f) => add_function_impl(acc, ctx, &f),
+                hir::AssocItem::TypeAlias(t) => add_type_alias_impl(acc, ctx, &t),
+                hir::AssocItem::Const(c) => add_const_impl(acc, ctx, &c),
+            }
+        }
+    }
+}
+
+fn add_function_impl(acc: &mut Completions, ctx: &CompletionContext, func: &hir::Function) {
+    use crate::display::FunctionSignature;
```

Comment:
We usually move all imports to the beginning of the file.

Reply:
Fixed.